### ISSUE TITLE
Q: Error read timeout

### ISSUE THREAD

**Issue by matteocrippa** [Orion98MC/node_apns#1]
Hi, today I received a couple of times this error meanwhile sending a batch list of push notifications.
The system stopped to send any notifications.

I wonder to ask you if you have any idea of the source of the problem, it would be related to appl servers or something different?

Do you have any trick/workaround to deal with this? On Error retry the push sending process?

**Comment by Orion98MC** [comment]
Hello Matteo, can you copy/past the stack trace if any ?

**Comment by matteocrippa** [comment]
the full message I receive is this:

Error! { [Error: read ETIMEDOUT] code: 'ETIMEDOUT', errno: 'ETIMEDOUT', syscall: 'read' }

**Comment by Orion98MC** [comment]
I assume you receive this in the on 'error' event of the push right? 
Is it happening after a long period of inactivity of the push connection? Can you reproduce it?
It may happen that the connection to apple's servers times out. Then when you try to send again, you may see this error. But it should be harmless since the connection should be rebuilt right after.

**Comment by matteocrippa** [comment]
Yep, the error is triggered by Error event.

I can give you these details, the app works in this way using these libs:
- superagent
- cron

The app, using the cron, cycle the code every 10 minutes.
Every 10 minutes sagent request for a remote page with a json with a list of events.
If the events is nearly to occur (30 minutes before) the system query the user list db and retrieve the list of the devices that have to be warned by a push.

So now I start up a simple loop in which I start creating pushes via node_push.
There are more or less 100 user per each loop, so in an asynch way I create all the pushes, but as soon as the console is able to show me the last one the Error I sent before is raised.

Sometimes it happens, sometimes no.

Is there any way to try increasing the timeout to try to fix? Or do you think is better to query again the pushes list and try to send them again?

**Comment by Orion98MC** [comment]
Matteo,

The error is stating that the connection did timeout. It may occur when the connection was left inactive a long time and was severed by the server.

If you run the sending process as a cron job, it is most likely that the connection is brand new and should not timeout. If it's a third party lib named cron, then I assume that the process (PID) is never restarted and that the timeout is indicating that the connection was closed on the server side. 

If your network connection to the internet is always online, I believe that this timeout should be pretty high and at least it should not happen within a 10 minutes window except if Apple changed their connection persistency policy. 

But anyways when the server closes the tcp connection node_apns should see a FIN packet and close it's connection such that next time it will rebuild it on demand. 

An other case of error which result in the connection being closed by the server is often a malformed token or notification. You should register for 'notificationError' to get its feedback.

So, basically to troubleshoot the problem, you may:
- register for notificationError on the push object
- check notifications validity each time you send one
- check for apple's feedback for long time no seen tokens
- use a brand new process every ten minutes (maybe using Unix cron jobs)
- in the error handler do the appropriate thing to resend not sent notifications

For more investigation please show me a bit of your code.
